File: frameworks/js/napi/wallpaperextensionability/wallpaper_extension_ability.js
/*
 * Copyright (c) 2021 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

let WindowManager = requireNapi('window');
let WindowName = 'wallpaper';
let windowType = 2000;
let windowsCreated = false;
let WIDTH = 480;
let HEIGHT = 960;

class WallpaperExtensionAbility {
  createWallpaperWin() {
    console.log(`${WindowName} createWallpaperWin`);

    WindowManager.create(this.context, WindowName, windowType).then((win) => {
      console.log(`${WindowName} wallpaperWindow`);
      this.wallpaperWindow = win;
      console.log(this.wallpaperWindow);
      this.wallpaperWindow.moveTo(0, 0).then(() => {
        this.wallpaperWindow.resetSize(WIDTH, HEIGHT).then(() => {
          console.log(`${WindowName} resetSize ${this.wallpaperURI}`);
          this.loadUiContent(this.wallpaperURI);
          console.log(`${WindowName} window created`);
          windowsCreated = true;
        });
      });
    }, (error) => {
      console.log(`${WindowName} window createFailed, error.code = ${error.code}`);
    });
  }

  onCreate(want) {
    console.log(`${WindowName} onWallpaperExtensionAbilityCreate`);
  }

  setUiContent(uri) {
    console.log(`${WindowName} setUiContent`);
    if (windowsCreated) {
      console.log(`${WindowName} loadUiContent`);
      loadUiContent(uri);
    } else {
      console.log(`${WindowName} save wallpaperURI`);
      this.wallpaperURI = uri;
    }
  }

  loadUiContent(uri) {
    console.log(`${WindowName} initUiContent ${uri}`);
    this.wallpaperWindow.loadContent(uri).then(() => {
      console.log(`${WindowName} loadContent`);
      this.wallpaperWindow.show().then(() => {
        console.log(`${WindowName} window is show`);
      });
    });
  }

  onWallpaperChange(wallpaperType) {
    console.log(`${WindowName} onWallpaperChange ${wallpaperType}`);
  }

  onDestroy() {
    console.log(`${WindowName} onWallpaperExtensionAbilityDestroy`);
  }
}

export default WallpaperExtensionAbility;

File: frameworks/kits/extension/src/wallpaper_extension_ability.cpp
/*
 * Copyright (c) 2021 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "wallpaper_extension_ability.h"

#include "connection_manager.h"
#include "hilog_wrapper.h"
#include "js_wallpaper_extension_ability.h"
#include "runtime.h"
#include "wallpaper_extension_context.h"

namespace OHOS {
namespace AbilityRuntime {
using namespace OHOS::AppExecFwk;
WallpaperExtensionAbility *WallpaperExtensionAbility::Create(const std::unique_ptr<Runtime> &runtime)
{
    HILOG_INFO("jws WallpaperExtensionAbility::Create runtime");
    if (!runtime) {
        return new WallpaperExtensionAbility();
    }
    HILOG_INFO("jws WallpaperExtensionAbility::Create runtime");
    switch (runtime->GetLanguage()) {
        case Runtime::Language::JS:
            return JsWallpaperExtensionAbility::Create(runtime);

        default:
            return new WallpaperExtensionAbility();
    }
}

void WallpaperExtensionAbility::Init(const std::shared_ptr<AbilityLocalRecord> &record,
    const std::shared_ptr<OHOSApplication> &application, std::shared_ptr<AbilityHandler> &handler,
    const sptr<IRemoteObject> &token)
{
    HILOG_INFO("jws WallpaperExtensionAbility begin init context");
    ExtensionBase<WallpaperExtensionContext>::Init(record, application, handler, token);
}

std::shared_ptr<WallpaperExtensionContext> WallpaperExtensionAbility::CreateAndInitContext(
    const std::shared_ptr<AbilityLocalRecord> &record,
    const std::shared_ptr<OHOSApplication> &application, std::shared_ptr<AbilityHandler> &handler,
    const sptr<IRemoteObject> &token)
{
    HILOG_INFO("jws WallpaperExtensionAbility begin init context");
    std::shared_ptr<WallpaperExtensionContext> context =
        ExtensionBase<WallpaperExtensionContext>::CreateAndInitContext(record, application, handler, token);
    if (context == nullptr) {
        HILOG_ERROR("WallpaperExtensionAbility::CreateAndInitContext context is nullptr");
        return context;
    }
    return context;
}
} // namespace AbilityRuntime
} // namespace OHOS